src/compiler/tsbuild.ts

```diff
@@ -0,0 +1,1222 @@
+                // Already visited
+                if (permanentMarks[projPath]) return;
+                // Circular
+                if (temporaryMarks[projPath]) {
+                    if (!inCircularContext) {
+                        hadError = true;
+                        reportDiagnostic(createCompilerDiagnostic(Diagnostics.Project_references_may_not_form_a_circular_graph_Cycle_detected_Colon_0, circularityReportStack.join("\r\n")));
+                        return;
+                    }
+                }
+
+                temporaryMarks[projPath] = true;
+                circularityReportStack.push(projPath);
+                const parsed = configFileCache.parseConfigFile(projPath);
+                if (parsed === undefined) {
+                    hadError = true;
+                    return;
+                }
+                if (parsed.projectReferences) {
+                    for (const ref of parsed.projectReferences) {
+                        const resolvedRefPath = resolveProjectName(ref.path);
+                        if (resolvedRefPath === undefined) {
+                            hadError = true;
+                            break;
+                        }
+                        visit(resolvedRefPath, inCircularContext || ref.circular);
+                        graph.addReference(projPath, resolvedRefPath);
+                    }
+                }
+
+                circularityReportStack.pop();
+                permanentMarks[projPath] = true;
+                buildOrder.push(projPath);
+            }
+        }
+
+        function buildSingleProject(proj: ResolvedConfigFileName): BuildResultFlags {
+            if (context.options.dry) {
+                reportDiagnostic(createCompilerDiagnostic(Diagnostics.A_non_dry_build_would_build_project_0, proj));
+                return BuildResultFlags.Success;
+            }
+
+            context.verbose(Diagnostics.Building_project_0, proj);
+
+            let resultFlags = BuildResultFlags.None;
+            resultFlags |= BuildResultFlags.DeclarationOutputUnchanged;
+
+            const configFile = configFileCache.parseConfigFile(proj);
+            if (!configFile) {
+                // Failed to read the config file
+                resultFlags |= BuildResultFlags.ConfigFileErrors;
+                context.projectStatus.setValue(proj, { type: UpToDateStatusType.Unbuildable, reason: "Config file errors" });
+                return resultFlags;
+            }
+
+            if (configFile.fileNames.length === 0) {
+                // Nothing to build - must be a solution file, basically
+                return BuildResultFlags.None;
+            }
+
+            const programOptions: CreateProgramOptions = {
+                projectReferences: configFile.projectReferences,
+                host,
+                rootNames: configFile.fileNames,
+                options: configFile.options
+            };
+            const program = createProgram(programOptions);

```

Comment:
should not this just call inot the same routine we have in tsc.ts to perform the compilation, why does it have to be different?